content/blog/2025-06-06-a-plan-for-simd.md

```diff
@@ -0,0 +1,142 @@
+This would operate in one of two modes, either as a proc macro run at build time, or as codegen to generate .rs files which would typically be checked into a repo, to minimize build time impact.
+(I’ll note that this is a particular instance of a much more general need for better staged compilation, which we’re feeling particularly acutely for shader compilation).
+At some point I'll post the prototype, as I think it's worthy of being considered if we're going to be doing a full exploration.
+
+While I think this approach has some advantages, it approaches the cost of building a real programming language, with associated needs for tooling etc.
+In addition, the proc macro approach really shows seams when it comes to cross-module interactions.
+
+Another possibility is to evolve pulp in the direction we need.
+That’s still not out of the question, but the changes proposed are quite extensive, and this could be disruptive to the existing user base, particularly faer.
+One goal in publishing this plan is to gather feedback from the pulp community about what they’d like to see happen.
+
+## On RVV and SVE
+
+One topic I didn’t cover in my blog post is RVV and SVE, which are pretty marginal these days but will become more important.
+There are some pretty big challenges, and for the most part we’re blocked on Rust support for the intrinsics.
+
+Taking a longer view, I think there are two modes for “scalable vector” SIMD.
+One is to generate asm which operates correctly no matter the SIMD width.
+It’s fairly easy to see how to do this for map-like workloads, and I think it’s reasonable to consider this mostly a codegen problem for autovectorization, as opposed to a good use case for explicit SIMD.
+
+The other is to treat scalable vectors as an annoying instruction encoding for SIMD width not known at compile time, and with partial support for operating on fixed-width (128 bit) blocks.
+This is essentially the same concept as “[Fixed Width Vector Workloads on Variable Width Vectors](https://gist.github.com/zingaburga/805669eb891c820bd220418ee3f0d6bd#fixed-width-vector-workloads-on-variable-width-vectors)”.
+The entire concept is a bit frustrating because nearly all ARM chips are 128 bits anyway.
+
+See also the paper [SIMD Everywhere Optimization from ARM NEON to RISC-V Vector Extensions](https://arxiv.org/pdf/2309.16509) for an approach to making explicit SIMD code portable to RVV.
+
+## Considerations for WASM SIMD
+
+There’s a [Zulip thread](https://xi.zulipchat.com/#narrow/channel/197075-vello/topic/WASM.20SIMD) on this.
+One tricky bit is that WASM doesn’t have runtime feature detection, rather they expect feature detection to be done as part of the negotiation for deciding which WASM blob to serve.
+In some ways, this makes sense, as it avoids pretty much all of the difficulties of multiversioning (including potential binary size impact), but on the other hand it’s a pain, as it complicates build and deployment.
```

Comment:
I think this is a less than generous characterization. Having shipped stuff in wasm land before, this was actually a nicer simplification as we were already doing this for various baselines of features so that we could take advantage of some other wasm features as they stabilized and shipped. It also let us do things like not serve something up to a browser that had a known issue without having to sort that out in the actual code within the blob.

I'd view this as a big simplification, not a complication or "tricky bit". :)